lib/active_model/serializer.rb

```diff
@@ -102,24 +102,49 @@ def self.attributes(*attrs)
       end
     end
 
+    # TODO: remove the dynamic method definition
     # @example
     #   class AdminAuthorSerializer < ActiveModel::Serializer
     #     attributes :id, :recent_edits
     #     attribute :name, key: :title
     #
+    #     attribute :full_name do
+    #       "#{object.first_name} #{object.last_name}"
+    #     end
+    #
     #     def recent_edits
     #       object.edits.last(5)
-    #     enr
-    def self.attribute(attr, options = {})
+    #     end
+    def self.attribute(attr, options = {}, &block)
       key = options.fetch(:key, attr)
-      _attributes_keys[attr] = { key: key } if key != attr
-      _attributes << key unless _attributes.include?(key)
+      reader = if block
+                 ->(instance) { instance.instance_eval(&block) }
+               elsif _fragmented
+                 ->(instance) { instance.class._fragmented.read_attribute_for_serialization(attr) }
+               else
+                 ->(instance) { instance.read_attribute_for_serialization(attr) }
+               end
```

Comment:
I'm not a fan of this style of if/else/end assignment. I think I'd rather a case statement. @noahsilas ?

There's a possible order-of-evaluation failure here if _fragmented isn't set before the attribute is defined. We'd have to move the logic into either `#attributes` or `#read_attribute_for_serialization` to make it class function order independent, but I think that just muddies up the code a bit